Revised training core

Drenderer opened the pull request:
### Key improvements:
- Split training ingredients into state (model and optimizer state) and static (optimizer, batcher, loss, etc.)
- Defined a new core `training_loop` function that only takes state, static, and callbacks as arguments. The loop updates the state.
- Removed `CallbackArgs` in favor of a `TrainingView`, which provides read-only access to static and read-write access to the state. This greatly simplifies the callbacks.
- Defined an ABC for Losses, which allows for custom gradient computations and implements model unwrapping by default (which avoids a redefinition of the loss in the fit function).
- Refactored datahandler.
- Removed the `HistoryCallback` in favor of a more general logging framework using a `MetricLogger` callback and a dict-like `History` object. This significantly improves the separation of concerns and enables the user to easily track custom metrics.
- Refactored `klax.fit` to recreate the old behaviour with the new components.

Drenderer committed:
Started a draft for the new trianing framework.

Drenderer committed:
Created fist working version of the new training core.

Drenderer committed:
Added simple benchmark script for measuring training time.

Drenderer committed:
Removed plotting from benchmark script.

Drenderer committed:
TrainingState pytree registration for flattening trick.

Drenderer committed:
Fixed issue with loss decorator and redefined mse and mae losses with the decorator.

Drenderer committed:
Improved comments, docstrings and HistoryCallback plotting behaviour.

Drenderer committed:
Updated the docs examples to work with the new training loop

Drenderer committed:
Jitted loss value and value_and_grad for speeding up training and
validation loss computations.

Drenderer committed:
Fixed some type annotations.

Drenderer committed:
Further typing fixes.

jaosch commented on klax/_losses.py:
```diff
@@ -12,36 +12,40 @@
 
     Example:
         A simple custom loss that computes the mean squared error between
         the predicted values `y_pred` and true values `y` for in inputs `x` may
```
Typo: delete "in"

jaosch commented on klax/_losses.py:
```diff
@@ -50,73 +54,112 @@ class Loss(Protocol):
-            data: The input data or structure used for loss computation.
-            batch_axis: Specifies the axis or axes corresponding to the batch
+            batch: The input data or structure used for loss computation.
+            batch_axes: Specifies the axis or axes corresponding to the batch
```
I don't fully agree with this description, since batch_axes could also be, e.g., an `eqx.Module`.

jaosch commented on klax/_losses.py:
```diff
@@ -50,73 +54,112 @@ class Loss(Protocol):
 
-mse = MSE()
 
+def loss(func: Callable) -> Loss:
```
This decorator is very cool

jaosch commented on klax/_training.py:
```diff
@@ -24,16 +25,57 @@
+
+    for step in range(1, static.steps + 1):
+        state, step_loss = make_step(state, next(static.batcher))
+        if any(
```
Insert the following comment above:

```
# Run all callbacks and break if any of them request termination of
# the training loop.
# Note! The square brackets are important. Otherwise the loop is
# terminated with the first callback that returns true. But we want
# to run all callbacks first and then decide, whether to terminate.
```

jaosch commented on klax/_training.py:
```diff
@@ -24,16 +25,57 @@
+from ._wrappers import apply
+
+
+def training_loop(
```
I suggest a name change to, e.g., `loop_training`, `loop_fit`, `fit_core`, `make_training_loop`, because a function should always start with a verb.

jaosch commented on klax/_training.py:
```diff
@@ -140,64 +185,6 @@ def fit[T: eqx.Module, H: Callback](
-    # that only leafs of type ArrayLike are vmapped. Thus it is possible to
-    # have data like `(str, array)` ans still use `batch_axis=0` instead of
-    # `batch_axis=(None, 0)`.
-    batch_axis, dataset_size = broadcast_and_get_size(data, batch_axis)
```
Happy to see this go :)

Drenderer replied:
Yeah! :D

jaosch commented on klax/_training.py:
```diff
@@ -210,53 +197,42 @@ def make_step(batch, flat_model, optimizer, flat_opt_state):
     # initially
     model = apply(model)
```
I agree that this should be called `rewarp` in the future.

Drenderer replied:
👍

jaosch commented on klax/_training.py:
```diff
@@ -210,53 +197,42 @@ def make_step(batch, flat_model, optimizer, flat_opt_state):
+                partial(loss.value, batch_axes=batch_axes),
+            )
+        history = HistoryCallback(
+            metric_defs=metric_defs,
```
Does this `metric_def` mean that there is no lazy evaluation of the loss anymore?

Drenderer replied:
Yes, and no. Klax no longer makes sure to cache loss itself. However, to make this cached just make sure that the metrics that you may compute multiple times each step are computed using a cached function, for example by decorating the function with @functools.cache.

jaosch commented on klax/_callbacks.py:
```diff
@@ -211,57 +103,73 @@ def __repr__(self):
+        """
+        if step % self.log_every == 0:
+            self.steps.append(self.step_offset + step)
+            self.metrics["step_loss"].append(step_loss)
```
I don't find the name `step_loss` very intuitive. I understand that there is one loss value for every recorded step, but calling it `step_loss` created more confusion than additional information in my opinion.

Drenderer replied:
batch_loss maybe?

Drenderer replied:
Or minibatch_loss.

jaosch commented on klax/_callbacks.py:
```diff
@@ -211,57 +103,73 @@ def __repr__(self):
+            self.steps.append(self.step_offset + step)
+            self.metrics["step_loss"].append(step_loss)
+            for name, (data, metric_fn) in self.metric_defs.items():
+                metric_value = metric_fn(state.model, data)
```
How would we include a cache here?

I think this could be accoplished by somthing like creating a `TrainingCache` class that gets the loss functions from as inputs and instead of passing the loss functions directly to the the history, we passe the cached versions.

Or maybe we can create a `@cache_training` decorator, that can can be wrapped around any function and stores it results in a global cache.

Drenderer replied:
Just chache the metric_fn, with metric_fn = functools.cache(metric_fn), and also use the cached version elsewhere.

jaosch reviewed (changes requested):
Very cool changes. The main challenge will be to implement a cache, that is easy to reuse for any other callback.

Drenderer committed:
Renamed step_loss to batch_loss and training_loop to run_training_loop, improved docs and removed the data from metric_defs, which now only require the model as input to compute the metrics.

Drenderer commented:
> Very cool changes. The main challenge will be to implement a cache, that is easy to reuse for any other callback.

Thanks! :)
But I disagree about the caching, the user should just provide a cached function as metric_def.

Drenderer commented:
With the history callback now storing the metric definitions it is no longer possible to pickle it. This makes saving difficult. It would be possible to just not save the metric_defs, but that means that after loading the user needs to supply the defs again if they want to continue training. 😐

Drenderer committed:
Some minor improvements and docs updates

Drenderer committed:
Started work on overhauling the tests.

jaosch commented:
> With the history callback now storing the metric definitions it is no longer possible to pickle it. This makes saving difficult. It would be possible to just not save the metric_defs, but that means that after loading the user needs to supply the defs again if they want to continue training. 😐

This is a similar problem to model serialization with hyperparameters (see [`https://docs.kidger.site/equinox/examples/serialisation/`]()). But I am not yet sure, what the solution could be.

jaosch committed:
Reverted change in broadcast_and_get_size that fails for nested PyTree.

The recent changes have been mostly reverted, because jax.tree.reduce
does not give the desired output with `operator.eq`.

jaosch committed:
Renamed batch_axis to batch_axes.

Datahandler tests now pass.

jaosch committed:
Fixed default value for total_step_digits.

jaosch committed:
Added a partitioned version to Loss

This enables compatibility with any
`optax.GradientTransormationExtraArgs` without any additional
boilerplate for the user.

jaosch committed:
Fixed `fit` and enabled arbitray optimizers in `run_training_loop`.

The `fit` function now calls the `TrainState.create` function for lack
of an `__init__`.

The `run_training_loop` function now uses a partitioned version of the
loss enable the use of arbitrary opatax optimizers again.

jaosch committed:
Comparibility with `optax.GradientTransformation

`TrainStatic` made compatible with `optax.GradientTransformation` by
casting them to `optay.GradientTransformationExtraArgs`. This required
the implementation of a constructor, because field converters are not
yet implemented in Python (see PEP 712)

jaosch committed:
Fixed callback tests except for saving and loading history.

jaosch committed:
Update training tests for new API

All tests in test_training.py pass now.

Drenderer committed:
Started work on a more strict distinction between static and state for the generalized training.

Drenderer committed:
Continued work on strict destinction between state and static.

Drenderer committed:
Removed HistoryCallback in favor of a MetricLogger and a History container. Implemented and integrated a MetricLogger. Streamlined training functionality.

Drenderer committed:
Added or modified tests for logging, losses, trainstate and training.

Drenderer committed:
Added an optional tqdm-based progressbar to the logger as new default, instead of spamming.

Drenderer committed:
Implemented combining histories via extension.

Drenderer committed:
Implemented saving and loading of history containers.

Drenderer committed:
Implemented basic plotting functionality for the history.

Drenderer committed:
Changed the callback call function to `on_training_step` and added a more comprehensive docstring.

Drenderer committed:
Removed unused import in _datahandler.py.

Drenderer committed:
Keys function for the history and improved docstrings.

Drenderer committed:
Major overhaul of the documentation.

Drenderer committed:
Updated simple training test script.

Drenderer committed:
Merge branch 'develop' into feature/generalized_training

Drenderer committed:
Fixed batch_axis -> batch_axes in docstring.

Drenderer committed:
Merge remote-tracking branch 'origin/develop' into feature/generalized_training

Drenderer committed:
Merge branch 'feature/generalized_training' into feature/progressbar

Drenderer committed:
Minor script change

Drenderer committed:
Merge branch 'develop' into feature/progressbar

Drenderer committed:
Fixed missing import in __init__.py

Drenderer committed:
Capitalized custom types in logging.

Drenderer committed:
Fixed logging test issue with capturing printed output.

Drenderer committed:
Merge pull request #82 from Drenderer/feature/progressbar

Drenderer committed:
Small update to the wrapper docs.

Drenderer committed:
Fixed issue with overwriting the default label in History.plot().

Drenderer committed:
Updated the examples to the new generalized training loop api.

Drenderer committed:
Refactored some scripts for the new training core.
